Refactor pkg manifest generator as interface

jberkhahn opened the pull request:
**Description of the change:**
Refactor the package manifest generator as part of an interface so it can be faked out in unit tests. A followup PR will refactor the cmd layer and hook it up to the new interface.

**Motivation for the change:**
See #3246. This is part of an ongoing effort to make OperatorSDK more test-friendly. This is the pattern I'd like to use for the rest of the generator package, and eventually most of the internal backend libs of OperatorSDK.

**Checklist**

If the pull request includes user-facing changes, extra documentation is required:
- [ ] Add a new changelog fragment in `changelog/fragments` (see [`changelog/fragments/00-template.yaml`](https://github.com/operator-framework/operator-sdk/tree/master/changelog/fragments/00-template.yaml))
- [ ] Add or update relevant sections of the docs website in [`website/content/en/docs`](https://github.com/operator-framework/operator-sdk/tree/master/website/content/en/docs)

camilamacedo86 commented on internal/generate/generate.go:
```diff
@@ -0,0 +1,26 @@
+// Copyright 2020 The Operator-SDK Authors
```
I see that these files are added as new. However, this implementation is not existent? 
So, shows that the project will kept both. IMO would be better here not keep the code duplicated in the repo and if required to do here already your idea for follow up indeed to make clear the motivations of the change.

jberkhahn replied:
I didn't want to include both changes in a single PR as the cmd bit is going to require some refactoring as well. I wanted to get peoples opinion on the structure of the new interface so I could change it now and not have to redo the cmd part.

estroz replied:
I would rather see these changes in cmd in this PR so we can comment on whether they make sense from the caller's perspective.

camilamacedo86 replied:
HI @jberkhahn,

The problem is how we can ` I wanted to get peoples opinion on the structure of the new interface` if we do not see what was changed/removed? Because of this, IMO it is hard to get your idea from here.

jberkhahn replied:
Ok, I can include the changes to cmd to show what it'll look like all together.

estroz commented on internal/generate/generate.go:
```diff
@@ -0,0 +1,26 @@
+// GeneratorSDK is an interface that specifies the Generate methods
+// to generate and write various OperatorSDK types such as
+// package manifests
+type GeneratorSDK interface {
+	//GeneratePackageManifest(opts *PkgOptions) error
+	GeneratePackageManifest(operatorName, version, outputDir string, opts ...*PkgOptions) error
+}
```
This interface setup seems to not follow reusable/extensible principles, since we'd have to add/remove methods whenever we add/remove a generator. I'd rather see something like
```go
type Generator interface {
	Generate(*Context) error
}

type Context struct {
	GetWriter func(string) (io.Writer, error)
	// Other fields used for writing
}
```

then we can do something like
```go
gen := NewPackageManifestGenerator(operatorName, version, outputDir string, opts ...*PkgOptions)
ctx := &Context{}
ctx.GetWriter = func(path string) {
	f, err := os.Open(path)
	return f, err
}
if err := gen.Generate(ctx); err != nil {
	return err
}
```

jberkhahn replied:
So, the issue there is the context would have to have all the info for all of the generate methods - which I think would be confusing. I also really don't like the callback method structure used in the original interface, and would much prefer to just have simple arguments passing the data into the generate method. (Think about how this will be tested in the faked out interface in the cmd layer. How am I going to check that the callback methods are set correctly? With an argument I can just make a simple assertion that the args are the correct values).

While the implementation I went with isn't the most minimalist, is it really that bad? How often will we be adding/removing generators? It seems like it's a write-once keep-forever kind of deal. I did stick the stuff that's likely to change (the optional args) into a struct, so the interface wont have to be changed and the fake won't have to be regenerated when we make changes.

joelanford replied:
What is this `Generator` interface is buying us? Wouldn't we always be creating and using the concrete type on the cmd side? Is there anywhere that we would actually use a generic `Generator` interface other than to write a high level test?

It's totally possible that I'm missing the context with this PR, but perhaps a more valuable way to test the generators would be to restructure the internals as interfaces and/or [overridable functions](https://github.com/operator-framework/operator-lib/blob/aac7eed1886158f2b42c11ade64b9e0673a5d3d6/leader/leader.go#L42-L44) such that we could [substitute different mock pieces in and out](https://github.com/operator-framework/operator-lib/blob/aac7eed1886158f2b42c11ade64b9e0673a5d3d6/leader/leader_test.go#L73-L75) to more easily exercise the real code paths?

IMO we should make the cmd side _extremely simple_, to the point that it wouldn't really be valuable to write tests for it.

And the actual implementation that's called by the cmd code is where we should refactor what we have into nice testable standalone libraries.

This is pretty much how I structured `kubectl-operator` (between [`internal/cmd`](https://github.com/operator-framework/kubectl-operator/tree/main/internal/cmd) and [`internal/pkg/action`](https://github.com/operator-framework/kubectl-operator/tree/main/internal/pkg/action)) though I haven't gotten around to writing tests for it yet.

camilamacedo86 replied:
I understand that the tool generates PkgManifest and Bundle. Am I right? 

Then, besides I am not very familiarized with this piece of code since I do not work on it. Shows that the appropriated design pattern that could be adopted to solve its is [Builder](https://golangbyexample.com/builder-pattern-golang/).

jberkhahn replied:
> restructure the internals as interfaces and/or [overridable functions](https://github.com/operator-framework/operator-lib/blob/aac7eed1886158f2b42c11ade64b9e0673a5d3d6/leader/leader.go#L42-L44) such that we could [substitute different mock pieces in and out](https://github.com/operator-framework/operator-lib/blob/aac7eed1886158f2b42c11ade64b9e0673a5d3d6/leader/leader_test.go#L73-L75) to more easily exercise the real code paths

This is what I am attempting to do by this PR. I wasn't including the cmd part just yet to keep things simple - I will add it on top of this PR.

camilamacedo86 replied:
Hi @jberkhahn Could we refractory it adopting a design pattern? In this case, I cannot see a better option than the [Builder](https://golangbyexample.com/builder-pattern-golang/). However, I am open to suggestions. 

I just think that we need to start looking for a better Maintainability.

jberkhahn replied:
I don't think Builder is a good fit for this particular bit. The cmd layer calls out to the generate package to generate a thing and write it to disk or stdout - it doesn't really care about getting the object back or even how it's constructed in the first place. I'm hoping my new implementation is much more maintainable. My intent is when a new generator is added to the interface, the method has required args as args and then a struct containing optional params. That way, if we need to change something or add functionality later, you just add a new field on the opts struct and change the implementation/tests. The interface doesn't have to change and the fake doesn't have to be regenerated, so again the only thing that needs to change in the next layer up is the implementation/tests.

camilamacedo86 replied:
I am maybe getting wrong but shows that your motivation here is to change the code for you be able to test it easily as you did for other commands and not to shape the solution base on what it does and represents. 

I think has a work in progress regards this which is related to add the `run bundle` and then, the next step after that would be cleanup/shape the code. 

So, your tests maybe would better fit after that. However, if @jmrodri @estroz which I think are working on with be ok with move forward with what you are proposing here and it will not affect the wip .. so I am fine with it as well.

camilamacedo86 commented on internal/generate/packagemanifest.go:
```diff
@@ -0,0 +1,180 @@
+		myOpts = opts[0]
+	}
+
+	pkg, err := g.generatePackageManifest(operatorName, version, myOpts)
```
IMO it shows strange. Why we have an exported and n internal func with the same name?

estroz replied:
This is a very common pattern.

jberkhahn replied:
that bit is copied from the original interface. GeneratePackageManifest is (generate a package manifest and all the other needed junk like write it to disk. generatePackageManifest is (actually generate me a package manifest object)

camilamacedo86 replied:
The original is not good and because of this, we are trying to refractory it. Am I right? So, let's try to shape it as better as possible. 

Regards the scenario, IMO it reduces the maintainability since we have an unambiguous representation which shows hurting principles such as DRY (Don’t Repeat Yourself).

However, I am human been and I can be wrong. Also, I love to learn something new. So, why it would be a good practice? When we should name 2 funcs with the same name? How it can help regards the readability? Could you please provide a reference that supports it as good practice? Have we any reason/argumentation that supports to adopt this practice at all?

jberkhahn replied:
generatePackageManifest is a helper method to help separate the different logical parts of GeneratePackageManifest to make it easier to reason about. It's a private method so it can't be used outside of this package. I don't think the naming is too confusing but am also not particularly worried about the change. My primary intent about this refactor was to make the relationship between the cmd layer and the generator more easily testable, and removing confusing indirection like the callback functions used in the previous implementation.

camilamacedo86 replied:
IMO this overload would be OK and is a common pattern when we want to provide an API less complexity to facilitate its usage and keep the complexity internally. However,  see: 

`GeneratePackageManifest(operatorName, version, outputDir string, opts ...*PkgOptions)` //  The Export API is more complexity 
`g.generatePackageManifest(operatorName, version, myOpts)`

And the actions that are done between the `GeneratePackageManifest` to `generatePackageManifest` are not all related to generation at all. See that it is doing validations as well which is hurting at least a few concepts such as cushion and single responsibilities.

jberkhahn replied:
So what do you propose? Move the bits that do validation and write the file to disk into separate helper methods? They would still only be called internally here so it would just be moving the code around. The validation that's happening is just checking that I have the correct info to perform the method call, so it's not going to get moved out of this layer. The writing file to disk could be extracted, but that seems to be needlessly adding complexity unless we need to write packagemanifests to disk in multiple places, which I don't think we do.

camilamacedo86 commented on internal/generate/packagemanifest.go:
```diff
@@ -0,0 +1,180 @@
+		base.DefaultChannelName = "alpha"
+	}
+
+	if err = validatePackageManifest(base); err != nil {
```
Have a func Generate that also validates shows hurting principles like single responsibility and cohesion.

jberkhahn commented:
@estroz @camilamacedo86 @joelanford

I've added the other two commits onto this branch. I also refactored what exactly was in which commit a bit.
1. Implements the new interface (and now also includes the packagemanfest base stuff that was previously in generate/packagemanifest/bases, but not actually used anywhere. I also added tests for the 1 public method from there, GetBase()). Also adds the counterfeiter fake.
2. switches the cmd layer over to using the new interface, and adds unit tests that utilize the fake.
3. removes the now unused code of the old implementation.

LMKWYT

camilamacedo86 commented on internal/generate/generate.go:
```diff
@@ -0,0 +1,26 @@
+// package manifests
+type GeneratorSDK interface {
+	//GeneratePackageManifest(opts *PkgOptions) error
+	GeneratePackageManifest(operatorName, version, outputDir string, opts ...*PkgOptions) error
```
Why we need `opts ...*PkgOptions` if in the code it is getting just the first value?

jberkhahn replied:
...*PkgOptions is a variadic argument, which means it can be any number of PkgOptions. This is just some syntactic sugar so you can provide PkgOptions or not when you call the method - it's purely optional. It's never intended to be called with more than one. I can remove this if you think I'm trying to be too clever.

estroz commented:
/assign

jberkhahn commented:
Rebased and refactored this based on #4074 

Hopefully I got it right, we'll see if CI blows up. Otherwise this should be ready for another round of review.

@estroz

estroz reviewed (commented):
How about a simplified interface and options setup like the following

In `internal/generate/generate.go`:
```go
type Generator interface {
	Generate() error
}
```

In `internal/generate/packagemanifest.go`:
```go
type NewGeneratorFunc func(string, []string) generator.Generator

type generator struct {
	operatorName string
	channels []string
}

func (g generator) Generate() error { ... }

func NewGenerator(operatorName string, channels []string) generator.Generator {
	return generator{
		operatorName: operatorName,
		channels: channels,
	}
}
```

In `internal/cmd/operator-sdk/generate/packagemanifests`:
```go
func NewCmd() *cobra.Command {
	return newCmd(genpkg.NewGenerator)
}

func newCmd(gf genpkg.NewGeneratorFunc) *cobra.Command {
	c := &packagemanifestsCmd{}
	cmd := &cobra.Command{
		...
		RunE: func(cmd *cobra.Command, args []string) error {
			if err := c.run(gf); err != nil {
				log.Fatalf("Error generating package manifests: %v", err)
			},
	}
	return cmd
}

func (c packagemanifestsCmd) run(gf genpkg.NewGeneratorFunc) error {
	g := gf(c.operatorName, c.channels)
	if err := g.Generate(); err != nil {
		return err
	}
	...
}
```

Then you'll be able to call `newCmd(testGF)` where `testGF` returns a test `generate.Generator`.

jberkhahn commented:
@estroz ok, but because your interface only contains a generic Generate method, I won't be able to use counterfeiter to create the fake I need for the unit test, as the salient bits would now be in the NewGenerator method. I'd have to hand-roll my own fakes for every generator.

estroz commented:
@jberkhahn is it possible to generate a counterfeit `NewGeneratorFunc` if you made it into an interface:
```go
// internal/cmd/operator-sdk/generate/packagemanifests/cmd_test.go

type NewGenFunc interface {
	NewGeneratorFunc(string, []string) generator.Generator
}
```

jberkhahn commented:
yes, but then every different generator would have to have the same constructor method - and I'd be back to passing in callback funcs to set any args that were actually different, and that would messy as hell to test. Getting rid of callback funcs was (imo) the single biggest improvement from my refactor.

estroz commented:
>then every different generator would have to have the same constructor method

Each generator would have its own constructor method. At some point you have to differentiate between generators; I'm explicitly making that happen in the constructor in my example rather than in the interface, which will break implementers every time the interface changes -> constructor choice is extensible. The interface is internal so your solution would be ok from a breakage standpoint, but it is kinda confusing and wouldn't really apply to other CLI setup code in the SDK, whereas a callback constructor could.

>Getting rid of callback funcs was (imo) the single biggest improvement from my refactor.

I assume you mean functional options. I agree those were not the best design choice. However my suggestion constitutes one callback vs. many, and is a well established construction pattern (ex. controller-runtime's [`ClientBuilder`](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/manager#ClientBuilder) and many manager [`Options`](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/manager#Options) fields).

jberkhahn commented:
>Each generator would have its own constructor method.
sorry, i mispoke, you're correct. i meant they would all use the same generate method, and thus have to pass along the needed info via the constructor (which is really just a callback func that's passed around in the cmd layer at that point). I don't like this because I would have to hand-roll my own fake constructor methods for the cmd tests, which was what I was trying to avoid. and those fakes would have to be manually rewritten everytime the passed args change. 

>which will break implementers every time the interface changes
because i put all the optional fields into a struct that's passed through, the interface (and thus regenerating the fakes, etc) would only change when a new generator is added or the required arguments for an existing interface change. If a new optional arg is added (because some new feature got put in), you would add it to the options struct, implement it, and add the unit tests in the generator lib and the cmd layer, like normal. i think it's easier to understand when all the data gets passed along at a single point, no?

estroz commented:
/retest

estroz commented:
@jberkhahn looks like you can't rerun a workflow created over a month ago :disappointed:. Rebase this onto master and you should be good to go.

jberkhahn commented:
@estroz updated

jberkhahn commented:
also, counterfeiter apparently has integration with go modules. i tagged the new interface so you can run 'go generate ./...' from the base dir and it will regenerate the fake. it doesn't look like we use go generate for anything else?

estroz commented on internal/generate/packagemanifest/packagemanifest.go:
```diff
@@ -17,37 +17,60 @@ package packagemanifest
+// to generate and write various OperatorSDK types such as
+// package manifests
+//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6 . PackageManifestGeneratorSDK
+type PackageManifestGeneratorSDK interface {
```
The package name already contains package manifest
```suggestion
type Generator interface {
```

estroz commented on internal/generate/packagemanifest/packagemanifest.go:
```diff
@@ -17,37 +17,60 @@ package packagemanifest
 )
 
+// PackageManifestGenerator is an implementation of the PackageManifestGeneratorSDK interface
+type PackageManifestGenerator struct{}
```
To align names/be idiomatic:
```suggestion
type generator struct{}

func NewGenerator() Generator {
	return generator{}
}
```

estroz commented on internal/generate/packagemanifest/packagemanifest.go:
```diff
@@ -17,37 +17,60 @@ package packagemanifest
-	OperatorName string
-	// Version is the version of the operator being updated.
-	Version string
+type PkgOptions struct {
```
```suggestion
type Options struct {
```

estroz commented on internal/generate/packagemanifest/packagemanifest.go:
```diff
@@ -17,37 +17,60 @@ package packagemanifest
+// package manifests
+//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6 . PackageManifestGeneratorSDK
+type PackageManifestGeneratorSDK interface {
+	Generate(operatorName, version, outputDir string, opts ...*PkgOptions) error
```
Options are always used by the callers as of now, so how about making them non-variadic?
```suggestion
	Generate(operatorName, version, outputDir string, opts Options) error
```

estroz commented on internal/generate/packagemanifest/packagemanifest.go:
```diff
@@ -56,91 +79,58 @@ type Generator struct {
-			return err
-		}
+func (g PackageManifestGenerator) Generate(operatorName, version, outputDir string, opts ...*PkgOptions) error {
+	if operatorName == "" {
+		return ErrNoOpName
+	} else if version == "" {
+		return ErrNoVersion
+	} else if outputDir == "" {
+		return ErrNoOutputDir
```
Nit: move these into their own conditions
```suggestion
	if operatorName == "" {
		return ErrNoOpName
	}
	if version == "" {
		return ErrNoVersion
	}
	if outputDir == "" {
		return ErrNoOutputDir
```

estroz commented on internal/generate/packagemanifest/packagemanifest.go:
```diff
@@ -154,25 +144,9 @@ func (g *Generator) generate() (*apimanifests.PackageManifest, error) {
 // makePkgManFileName will return the file name of a PackageManifest.
 func makePkgManFileName(operatorName string) string {
-	return operatorName + packageManifestFileExt
+	return strings.ToLower(operatorName) + packageManifestFileExt
```
This is a breaking change unfortunately.

jberkhahn replied:
is it? it's been a while since i wrote that bit but I seem to remember i had to change it to keep up with established behavior in the e2e tests somewhere, let me look around

jberkhahn replied:
so, it looks like this field is subject to RFC 1123 as it's used for the name of kube objects elsewhere. If you try to run this on master right now with a PROJECT file that somehow has a capital project name or manually set it, operator-sdk blows up. With this change, operator-SDK still blows up in the same place. So this line should never actually do anything in actual use, it's just internal hardening. I suppose silently changing it is a bad idea though. Would checking if this field looks right and returning an error if it doesn't be OK?

estroz replied:
What do you mean by "blows up"? What if you set `generate packagemanifests --package Foo`?

This filename does not have to follow RFC 1123 because it is a filename. RFC 1123 applies to object names though.

jberkhahn replied:
on current master:
```
jberkhahn@Purgatory> operator-sdk generate packagemanifests --version 1.0.0 --deploy-dir foo --crds-dir api/v1/ --package FOOBAR
Generating package manifests version 1.0.0
ERRO[0000] ClusterServiceVersion validation: [CSVFileNotValid] (FOOBAR.v1.0.0) metadata.name "FOOBAR.v1.0.0" is invalid:
[a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')] 
FATA[0000] Error generating package manifests: error generating ClusterServiceVersion: invalid generated ClusterServiceVersion 
```

estroz replied:
Ah then perhaps there's no way for a package manifest file name to contain upper case characters because of preceding checks. In that case, lowercase the operator name.

jberkhahn replied:
after thinking about this some more, i think the real place to fix this is in the validator in operator-framework/api. removing it from this PR

estroz commented on internal/generate/packagemanifest/packagemanifest.go:
```diff
@@ -17,37 +17,60 @@ package packagemanifest
+}
+
+// PackageManifest configures the PackageManifest that GetBase() returns.
+type PackageManifest struct {
```
This no longer needs to be exported since it is only used within one package. In fact, you can make this into a function `func getBase(packageName, basePath string) (*apimanifests.PackageManifest, error)`

estroz reviewed (changes requested):
Looking good!

estroz commented on internal/cmd/operator-sdk/generate/packagemanifests/packagemanifests_suite_test.go:
```diff
@@ -0,0 +1,27 @@
+// Copyright 2020 The Operator-SDK Authors
```
```suggestion
// Copyright 2021 The Operator-SDK Authors
```

estroz commented on internal/cmd/operator-sdk/generate/packagemanifests/packagemanifests_test.go:
```diff
@@ -0,0 +1,262 @@
+// Copyright 2020 The Operator-SDK Authors
```
```suggestion
// Copyright 2021 The Operator-SDK Authors
```

estroz commented on internal/cmd/operator-sdk/generate/packagemanifests/packagemanifests_test.go:
```diff
@@ -0,0 +1,262 @@
+		})
+		Context("a valid project file is present", func() {
+			BeforeEach(func() {
+				err := os.Chdir("../../../../../testdata/go/v3/memcached-operator")
```
This is kinda a PITA but paths should be platform-agnosic.

estroz commented on internal/generate/packagemanifest/packagemanifest_suite_test.go:
```diff
@@ -0,0 +1,27 @@
+// Copyright 2020 The Operator-SDK Authors
```
```suggestion
// Copyright 2021 The Operator-SDK Authors
```

estroz commented on internal/generate/packagemanifest/packagemanifest.go:
```diff
@@ -17,37 +17,63 @@ package packagemanifest
 	"path/filepath"
 	"sort"
 
+	yaml "gopkg.in/yaml.v1"
+
```
```suggestion
	"sigs.k8s.io/yaml"
```

estroz commented on internal/cmd/operator-sdk/generate/packagemanifests/packagemanifests_test.go:
```diff
@@ -0,0 +1,262 @@
+	)
+	BeforeEach(func() {
+		c = packagemanifestsCmd{}
+		crdsDir = "crds/"
+		deployDir = "deploy/"
+		inputDir = "input/"
+		kustomizeDir = "kustomize/"
```
```suggestion
		crdsDir = "crds"
		deployDir = "deploy"
		inputDir = "input"
		kustomizeDir = "kustomize"
```

estroz commented on internal/cmd/operator-sdk/generate/packagemanifests/cmd.go:
```diff
@@ -18,6 +18,7 @@ import (
 	"fmt"
 	"path/filepath"
 
+	"github.com/operator-framework/operator-sdk/internal/generate/packagemanifest"
 	log "github.com/sirupsen/logrus"
 	"github.com/spf13/cobra"
 	"github.com/spf13/pflag"
```
```suggestion
	log "github.com/sirupsen/logrus"
	"github.com/spf13/cobra"
	"github.com/spf13/pflag"

	"github.com/operator-framework/operator-sdk/internal/generate/packagemanifest"
```

jberkhahn commented:
@estroz updated
fixed:
- copyright dates
- imports
- os-depdendant filepaths. i'm not 100% sure on this, is ".." one directory up in windows?
- removed the lowercase thing because I don't think this is the right place to fix it

estroz commented on internal/cmd/operator-sdk/generate/packagemanifests/cmd.go:
```diff
@@ -1,4 +1,4 @@
-// Copyright 2020 The Operator-SDK Authors
+// Copyright 2021 The Operator-SDK Authors
```
Only newly added files should have 2021 license headers.
```suggestion
// Copyright 2020 The Operator-SDK Authors
```

estroz commented on internal/generate/packagemanifest/packagemanifest.go:
```diff
@@ -17,37 +17,63 @@ package packagemanifest
+// to generate and write various package manifests
+//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6 . Generator
+type Generator interface {
+	Generate(operatorName, version, outputDir string, opts *Options) error
```
This shouldn't be a pointer (matter of preference though):
```suggestion
	Generate(operatorName, version, outputDir string, opts Options) error
```

jberkhahn replied:
why not? I made it a pointer intentionally so the opts struct (which could potentially grow to be quite large) doesn't have to get shuttled around everywhere

estroz replied:
I doubt the struct's fields + reference overhead will be > than 1kb, which is not worth fretting over for 2-3 copies. There's also a good chance the compiler will optimize parts of this movement. With pointers, you have to perform a `nil` check, which is an extra jump instruction and more LOC. Perhaps the compiler will optimize this too. I guess my point is that using a pointer makes just as much sense from a performance perspective as pass-by-value, so it comes down to code verbosity and potential nil pointer panics. Since `Options` is a required field, i.e. `opts != nil` will always be true as the code is written now, it makes sense to not use a pointer.

jberkhahn commented:
linter complained about the unexported generator type, was making the sanity tests fail so i changed it back

estroz commented on internal/generate/packagemanifest/packagemanifest.go:
```diff
@@ -17,37 +17,63 @@ package packagemanifest
 	packageManifestFileExt = ".package.yaml"
 )
 
+// GPkgManenerator is an implementation of the Generator interface
+type PkgManGenerator struct{}
```
This shouldn't be exported because it isn't used anywhere
```suggestion
// generator implements the Generator interface
type generator struct{}
```

estroz commented on internal/generate/packagemanifest/packagemanifest.go:
```diff
@@ -17,37 +17,63 @@ package packagemanifest
+// GPkgManenerator is an implementation of the Generator interface
+type PkgManGenerator struct{}
+
+// NewGenerator returns a new generator object
+func NewGenerator() PkgManGenerator {
+	return PkgManGenerator{}
+}
```
```suggestion
// NewGenerator returns a package manifest Generator.
func NewGenerator() Generator {
	return generator{}
}
```

jberkhahn replied:
oh, the more you know ---x

jberkhahn commented:
anything else on this? would like to make it in this release

estroz commented on internal/generate/packagemanifest/packagemanifest.go:
```diff
@@ -17,37 +17,63 @@ package packagemanifest
+
+// Generator is an interface that specifies the Generate methods
+// to generate and write various package manifests
+//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6 . Generator
```
Can you add `go generate ./...` to the `generate` make target? That way all counterfeits will always be regenerated.

estroz commented on internal/generate/packagemanifest/packagemanifest.go:
```diff
@@ -17,37 +17,63 @@ package packagemanifest
 	packageManifestFileExt = ".package.yaml"
 )
 
+// GPkgManenerator is an implementation of the Generator interface
```
```suggestion
// generator is an implementation of the Generator interface
```

estroz commented on internal/generate/packagemanifest/packagemanifest.go:
```diff
@@ -17,37 +17,63 @@ package packagemanifest
 	"sort"
 
+	"sigs.k8s.io/yaml"
+
```
```suggestion
```

estroz reviewed (approved):
Some nits and a suggestion for a nice-to-have.

jberkhahn commented:
@estroz updated

estroz reviewed (approved):
/lgtm

estroz commented:
@jberkhahn try running `make generate` then push `go.sum` changes.

jberkhahn commented:
@estroz hopefully that fixes it?
